### PR TITLE
Remove rpc.jsonrpc.method, clarify rpc.method instead.

### PR DESCRIPTION
**Pull request by Oberon00**
(no corresponding issue)

## Changes

Remove `rpc.jsonrpc.method`, clarify that `rpc.method` should be used instead.

## Related

* PR that introduced JSON RPC conventions: #1643 (@Rast1234)

This change was just released in 1.4, but semantic conventions are still experimental thus a breaking change like this may be acceptable.

### CODE REVIEW

**Comment by Rast1234** [comment]
So you changed the purpose of `rpc.method` to be logical, not physical method/function name on the server? Great!

**Comment by Oberon00** [comment]
> So you changed the purpose of rpc.method to be logical, not physical method/function name on the server

In my mind, this was already the case before, but the wording was not very clear.

**Comment by carlosalberto** [comment]
@Rast1234 Just to double confirm: are you ok with this change?

**Comment by Rast1234** [comment]
Yes!

File: semantic_conventions/trace/rpc.yaml [hunk 1]
```diff
@@ -108,3 +112,5 @@ groups:
           type: string
           brief: "`error.message` property of response if it is an error response."
           examples: ['Parse error', 'User already exists']
+      constraints:
+        - any_of: ['rpc.method']
```
Comment (arminru) [inline semantic_conventions/trace/rpc.yaml:116]: I think you could also include/reference it and make it required (unconditionally). The current markdown output of this still looks a bit odd.

File: semantic_conventions/trace/rpc.yaml [hunk 2]
```diff
@@ -12,13 +12,22 @@ groups:
           type: string
           required:
             conditional: "No, but recommended"
-          brief: 'The full name of the service being called, including its package name, if applicable.'
+          brief: 'The full (logical) name of the service being called, including its package name, if applicable.'
+          note: >
+            This is the logical name of the service from the RPC interface perspective,
+            which can be different from the name of any implementing class.
+            The `code.namespace` attribute may be used to store the latter
```
Comment (arminru) [inline semantic_conventions/trace/rpc.yaml:19]:
```suggestion
            The `code.namespace` attribute may be used on the server to store the latter
```

File: semantic_conventions/trace/rpc.yaml [hunk 3]
```diff
@@ -12,13 +12,22 @@ groups:
...
         - id: method
           type: string
           required:
             conditional: "No, but recommended"
-          brief: 'The name of the method being called, must be equal to the $method part in the span name.'
+          brief: 'The name of the (logical) method being called, must be equal to the $method part in the span name.'
+          note: >
+            This is the logical name of the method from the RPC interface perspective,
+            which can be different from the name of any implementing method/function.
+            The `code.function` attribute may be used to store the latter.
```
Comment (arminru) [inline semantic_conventions/trace/rpc.yaml:30]:
```suggestion
            The `code.function` attribute may be used on the server to store the latter.
```

**Review by arminru — APPROVED** [review]
Thanks! I think explicitly declaring this as the logical names and referencing the `code.*` attributes makes sense in any case, not only for JSON-RPC.

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
LGTM once the feedback has been addressed.

**Review by SergeyKanzhelev — APPROVED** [review]

File: semantic_conventions/trace/rpc.yaml (hunk 2 above)
Comment (Oberon00) [reply semantic_conventions/trace/rpc.yaml:19]: Hmm, the implementing method on the client *can* make sense, as many RPC frameworks generate client stub classes for you. Then you would store the name of the client stub class at the client side. Could use extra wording though. WDYT?

File: semantic_conventions/trace/rpc.yaml (hunk 1 above)
Comment (Oberon00) [reply semantic_conventions/trace/rpc.yaml:116]: Yeah, there is https://github.com/open-telemetry/build-tools/issues/33 for a related case.

**Comment by Oberon00** [comment]
@arminru
> makes sense in any case, not only for JSON-RPC.
Not sure if this was meant as a suggestion or statement, but that's what this PR does already?

File: semantic_conventions/trace/rpc.yaml (hunk 2 above)
Comment (arminru) [reply semantic_conventions/trace/rpc.yaml:19]: Sounds reasonable, let's go with suggesting both then.

File: semantic_conventions/trace/rpc.yaml (hunk 1 above)
Comment (Oberon00) [reply semantic_conventions/trace/rpc.yaml]: b6fd5ce

File: semantic_conventions/trace/rpc.yaml (hunk 2 above)
Comment (Oberon00) [reply semantic_conventions/trace/rpc.yaml:19]: ce3512b

File: semantic_conventions/trace/rpc.yaml (hunk 3 above)
Comment (Oberon00) [reply semantic_conventions/trace/rpc.yaml]: ce3512b

**Comment by arminru** [comment]
> > makes sense in any case, not only for JSON-RPC.
> 
> Not sure if this was meant as a suggestion or statement, but that's what this PR does already?

@Oberon00 Oh I was just saying that I like the change and that I agree refining the semantics beyond just solving it for JSON-RPC (as done in the PR) makes sense.

**Review by arminru — APPROVED** [review]

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
@Oberon00 Please update your PR so we can merge it.

**Comment by jkwatson** [comment]
This should include an update to the schema, correct? Our schema does support re-writing attributes from old to new, I believe.

**Comment by Oberon00** [comment]
@jkwatson 

> This should include an update to the schema, correct? Our schema does support re-writing attributes from old to new, I believe.

Hmm, probably, unless we consider schema updates part of the work of cutting a release and do it all at once at the end.
In that case we rename an attribute to an already existing attribute. I'm not sure what should happen if the existing rpc.method attribute is already filled out. Based on what I think was the misunderstanding about rpc.method that caused rpc.jsonrpc.method, I think that rpc.method should be overwritten with rpc.jsonrpc.method.

That being said, I would appreciate help with writing the schema, as there are no examples to draw from yet? @tigrannajaryan, can you assist?

**Comment by carlosalberto** [comment]
Ping @tigrannajaryan

**Comment by tigrannajaryan** [comment]
Assuming this is really semantically an attribute rename this schema file should do it:
```yaml
file_format: 1.0.0
schema_url: https://opentelemetry.io/schemas/1.5.0
versions:
  1.5.0:
    spans:
      changes:
        # Sequence of translations to apply to convert the schema from a prior version
        # to this version. The order in this sequence is important. Translations are
        # applied from top to bottom in the listed order.
        - rename_attributes:
            # Rename attributes of all spans, regardless of span name.
            # The keys are the old attribute name used prior to this version, the values are
            # the new attribute name starting from this version.
            attribute_map:
              rpc.jsonrpc.method: rpc.method
```
The file should be named `schemas/1.5.0` (no extension in the file name).

Note, I did not review the PR thoroughly, so I am not completely sure it is actually a renaming and not a deletion of an old attribute and introduction of a new unrelated attribute.

P.S. I have put together a small tool to do some sanity checks of schema files: https://github.com/tigrannajaryan/telemetry-schema/blob/main/cmd/scheck/main.go

**Comment by Oberon00** [comment]
> Note, I did not review the PR thoroughly, so I am not completely sure it is actually a renaming and not a deletion of an old attribute and introduction of a new unrelated attribute.

This PR is a deletion of a new attribute because it can already by represented by an existing attribute (new -> existing). So it's not exactly a renaming (new1 -> new2)

**Comment by tigrannajaryan** [comment]
> What would `rename_attributes` do if both the new and old attribute are set?

The behavior is undefined in the OTEP right now. We can define it but I would say if there is an expectation that both may be set then it does not sound like renaming and seems to be something more complicated. I don't think `rename_attributes` can handle that.

**Comment by tigrannajaryan** [comment]
> So then is it fine to merge this PR without a schema entry?

Yes, I think so. Since the semconv is not yet stable it is fine. In the future if something similar to this happens we will need to figure out how we can describe the change in the schema file.

**Pull request merged**